Implement tooling to generate multiple plotly schemas

PatrickVienne opened the pull request.

MetalBlueberry commented on generator/typefile.go:
```diff
@@ -81,6 +81,7 @@ func (file *typeFile) parseAttributes(namePrefix string, typePrefix string, attr
 	fields := make([]structField, 0, len(attr))
 
 	for _, name := range sortKeys(attr) {
+		//fmt.Println("parsing: ", namePrefix, typePrefix, name)
```
Do not leave commented lines in the code base

PatrickVienne replied:
thanks, removed

MetalBlueberry commented on generator/cmd/generator/main.go:
```diff
@@ -18,7 +19,7 @@ func (c Creator) Create(name string) (io.WriteCloser, error) {
 	schema := flag.String("schema", "schema.json", "plotly schema")
 	outputDirectory := flag.String("output-directory", "gen/", "output directory, must exist before generation")
-
+	*schema = "plotly-schema_v2.29.0.json"
```
Why is this hardcoded?

PatrickVienne replied:
Thank you, good catch, initially experimented with different versions and kept those files with the version name around.
Removed this one.

MetalBlueberry commented on README.md:
```diff
@@ -113,6 +110,35 @@ For strings... This is a little bit more complicated, In AWS package they are us
+Save this file as your new schema file and use it as input to regenerate your library.
+The library should be in graph_objects to overwrite the old types, and to be in the same folder as plotly.go.
+
+To solve multiple issues with generating invalid go files due to line-comments being too long, many descriptions
+are also shortened.
```
I do not understand this comment, I've never seen problems due to long comment lines, neither I can see the place where such descriptions are shortened. 
Can you provide any example?

PatrickVienne replied:
true, it got solved.
I had a malformed schema file first with line breaks in the description entries which then generated multiline text in the golang files where only the first line was commented out.

I'll remove this comment

MetalBlueberry commented on generator/renderer.go:
```diff
@@ -56,7 +58,26 @@ func (r *Renderer) CreateTrace(dir string, name string) error {
 	if err != nil {
-		return fmt.Errorf("cannot format source, %w", err)
+		switch err.(type) {
+		case scanner.ErrorList:
+			errorfile := "__tmp_error__.go"
+			fmt.Println("Bad Source:", dir, name)
+			fmt.Println(src.String())
+			fmt.Println("======")
+			for i, e := range err.(scanner.ErrorList) {
+				fmt.Printf("%d) cannot format source, %+v. %s\nadded error file in: %s", i, e, e.Error(), errorfile)
+			}
+			create, _ := os.Create(errorfile)
+			create.Write(src.Bytes())
+			create.Close()
+
+			if err != nil {
+				return err
+			}
+			return err
```
Why do we need this error handling? 
It feels more complex than it should be.

Also it creates a file without checking for error, which means this code can lead to a panic situation. 

Can we keep all error reporting just printing to the stdout?

PatrickVienne replied:
yes, this was helping me to identify build issues, but I'll remove the file creation

MetalBlueberry commented on README.md:
```diff
@@ -1,6 +1,3 @@
-[![Go Report Card](https://goreportcard.com/badge/github.com/MetalBlueberry/go-plotly)](https://goreportcard.com/report/github.com/MetalBlueberry/go-plotly)
-[![godoc](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/mod/github.com/MetalBlueberry/go-plotly?tab=overview)
-
```
Why is this removed?

PatrickVienne replied:
I put it back again.
I removed it as I could not use it on my fork, but since I'm merging back I added it again.

MetalBlueberry reviewed (changes requested):
Thank you very much for the contribution!! :heart: 

I've left a few comments with some questions :smile:

PatrickVienne commented:
@MetalBlueberry 
thanks for the comments, back to you.

MetalBlueberry commented:
Okey, that last commit is an attempt to reduce the number of changes on this PR.

PatrickVienne commented:
Hi @MetalBlueberry 

it is not generated, because it is not part of the json schema:

- go.mod
-> Agreed

- add instructioons:
-> 

These are the new trace types:
`jq '.schema.traces | keys' schema.json --sort-keys`
```
[
  "bar",
  "barpolar",
  "box",
  "candlestick",
  "carpet",
  "choropleth",
  "choroplethmapbox",
  "cone",
  "contour",
  "contourcarpet",
  "densitymapbox",
  "funnel",
  "funnelarea",
  "heatmap",
  "heatmapgl",
  "histogram",
  "histogram2d",
  "histogram2dcontour",
  "icicle",
  "image",
  "indicator",
  "isosurface",
  "mesh3d",
  "ohlc",
  "parcats",
  "parcoords",
  "pie",
  "pointcloud",
  "sankey",
  "scatter",
  "scatter3d",
  "scattercarpet",
  "scattergeo",
  "scattergl",
  "scattermapbox",
  "scatterpolar",
  "scatterpolargl",
  "scattersmith",
  "scatterternary",
  "splom",
  "streamtube",
  "sunburst",
  "surface",
  "table",
  "treemap",
  "violin",
  "volume",
  "waterfall"
]
```
Or via jsonviewer:
https://jsonviewer.stack.hu/
![image](https://github.com/MetalBlueberry/go-plotly/assets/21186220/3a1ef230-8b40-4f9f-9b61-c4ae0e92358d)

If this update goes through, I'll submit the next update to the latest version, watching out for your current comments. I also plan to add dark theme, as in plotly for python

I added a downlader script a description how to generate the new files without manuel work, only using the scripts.
Also added some more hints in the readme on using the scripts and tools like jq and jsonviewer.

Hope this helps to get it through the finish line :)
Thanks!

MetalBlueberry commented:
Alright! I added a few steps to the CI to automatically check that the scheme is correctly built and I've identified some bugs with consistency on the scheme generation :sweat_smile: 

Not perfect yet, But I've merged those changes here so we can get better feedback on your changes.

PatrickVienne commented:
Thanks a lot @MetalBlueberry 
anything I can do in support?

MetalBlueberry commented on generator/renderer.go:
```diff
@@ -466,6 +466,7 @@ var symbolMap = []string{
 	"+", "Plus",
 	"?", "Question",
 	"$", "Dollar",
+	".", "Dot",
```
BTW if you open a new PR with only this fix, and the comment aboive.  I will approve it right the way.

MetalBlueberry commented on generator/cmd/generator/main.go:
```diff
@@ -54,7 +54,7 @@ func main() {
 	}
 
-	err = os.MkdirAll(output, os.ModePerm)
+	err = os.MkdirAll(output, 0755)
```
This is also a good candidate for a single PR

PatrickVienne commented:
@MetalBlueberry 

great idea. Just slightly adjusted the yaml.

Hope this captures the essence of what you meant

running the downloader will download all the schema files as defined in the yaml.

running the generator still needs to be done for each version one by one, but can be done based on the downloaded files in the folder.

![image](https://github.com/MetalBlueberry/go-plotly/assets/21186220/e565dac1-8404-44a6-b3d1-24b8ad8b54d9)

slightly changed the schema yaml to make the versions more in line with the git tags of the plotly repo
![image](https://github.com/MetalBlueberry/go-plotly/assets/21186220/473768eb-8e94-4fb0-b003-d4c2ee8a4e9c)

won't have time for the next weeks to finish it unfortunately, but maybe this can get it started.

open todos still:
- adjust the readme file
- also create the plot.go file for separate versions if needed
- integrate CI generation of newly added versions
- read the yaml also in the generator, and use the output path of schemas.json to determine the output path.

PatrickVienne commented:
@MetalBlueberry 
completed all the points, and the generation now is also running in the workflow which you setup for testing

Also updated the readme now and added some release notes.

```
### Add version v2.29.1 and v2.31.1
- Added Downloader for schema file
- added generator for different packages based on plotly version in subfolders for each version
- enable easier use for different plotly version by new import paths which include the version generated. Example: `grob "github.com/MetalBlueberry/go-plotly/generated/v2.31.1/graph_objects"`
- removed go generate directive from the templates which are no longer necessary
- removed offline package, as each version package needs its own **plot.go**.
- Replaced all `offline` imports in the examples packages by just calling the `grob` package of the specified version
- all html's will now refer to the correct plotly version's cdn in the html's head
```

We've come a long way, thanks for your comments, really cool stuff, hope we can get this merged. and have next steps implemented in nxt PRs soon. 
Animations Support and Dark Theme are on my list.

![image](https://github.com/MetalBlueberry/go-plotly/assets/21186220/2010965f-2438-44c6-97fd-1c49eb16b45a)

PatrickVienne commented:
@MetalBlueberry 
I can't get the wasm example to run though.

one would also need to make sure to use the correct wasm_exec.js
https://github.com/golang/go/blob/go1.21.0/misc/wasm/wasm_exec.js

but this again is a topic for another PR.

PatrickVienne commented:
fixed another missing feature now, which prevented users from using the "arrayOk" feature.


![image](https://github.com/MetalBlueberry/go-plotly/assets/21186220/168353eb-e017-4717-b6d4-8dae0e2d3aed)

```go
		attr := attr[name]
		// if arrays are allowed, example marker sizes, then javascript allows multiple different inputs hard to map in golang:
		// [100,50,25,25,50,100] -> use different sizes for the markers
		// [100] -> only the first marker will have size 100, all others have 0 size
		// 100 -> all markers have size 100
		// this can only be captures by using interface and allowing the user to define the used type themselves.
		// this sadly means that there is no more type checking on those entries
		var ajust = func(typename string) string { return typename }
		if attr.ArrayOK {
			ajust = func(typename string) string { return "interface{}" }
		}
```

MetalBlueberry commented on generator/typefile.go:
```diff
@@ -86,13 +86,23 @@ func (file *typeFile) parseAttributes(namePrefix string, typePrefix string, attr
 		}
 
 		attr := attr[name]
+		// if arrays are allowed, example marker sizes, then javascript allows multiple different inputs hard to map in golang:
+		// [100,50,25,25,50,100] -> use different sizes for the markers
+		// [100] -> only the first marker will have size 100, all others have 0 size
+		// 100 -> all markers have size 100
+		// this can only be captures by using interface and allowing the user to define the used type themselves.
+		// this sadly means that there is no more type checking on those entries
+		var ajust = func(typename string) string { return typename }
+		if attr.ArrayOK {
+			ajust = func(typename string) string { return "interface{}" }
```
I understand this, but it would be nice to keep the scope of this PR to create multiple versions. We can merge this later :blush:

MetalBlueberry reviewed (changes requested):
First of all, Thank you very much :bow:  
I'm very happy to see your contributions and I'm impressed with your energy to contribute. 


Said that, there are a few things we have to do before this merge.

## Keep the offline package

The offline package was designed following the python interface. like you can see [here](https://stackoverflow.com/questions/34957790/use-plotly-offline-to-generate-graphs-as-images). This is important as it allows people to have a similar interface on multipla languages.

Also, the grob package is completely independent of the offline tools. So it doesn't make sense to keep in in the same directory.

I think you did this because now we have multiple plotly versions, so you actually need an offline package per version. IF that is the case, I believe we still have to keep the offline package on its own directory so it can be imported on demand. 

For example, the path for the offline package may be `generated/2.29/offline`

What do you think?

## Cleanup

We provably don't need the old graph_objects and offline packages. Do you have any reason to keep it? 

## more...

I did not have enough time to properly review this, I'm sorry. But I hope you have enough gudance to keep moving.

overall, I suggest you to try to keep the changes to the minimum. I know it is easy to get distracted and fix things along the way, but it makes it very hard for me to isolate changes and approve the pull request. Remember you don't have to fix everything now, you just want the minimum amount of changes that actually work and that people will benefit from. 

Thank you again :heart:

PatrickVienne commented:
@MetalBlueberry 
back to you

MetalBlueberry commented on generator/templates/plotly.tmpl:
```diff
@@ -21,6 +18,12 @@ type Trace interface {
 // Traces is a slice of Traces
 type Traces []Trace
 
+type Frame struct {
+	Name   string      `json:"name"`
+	Data   []Trace     `json:"data"`
+	Layout interface{} `json:"layout"` //'Layout properties which this frame modifies. The format is identical to the normal layout definition.'
+}
+
```
Same comment as before! Let's leave the frames for a different PR.

MetalBlueberry commented on generator/cmd/generator/main.go:
```diff
@@ -21,20 +45,34 @@ func (c Creator) Create(name string) (io.WriteCloser, error) {
+// FindDir goes back in the project directory to find the correct location of the schema and output folders
+// this allows the script to run smoothlessly with go generate in the cmd files,
+// and also when it is run from the project root as working directory
+func FindDir(schema, output string) (string, string) {
+	for i := 0; i < 4; i++ {
+		if _, err := os.Stat(schema); err == nil {
+			break
+		}
+		schema = filepath.Join("../", schema)
+		output = filepath.Join("../", output)
+	}
+	return schema, output
+}
```
This is risky, if there is a bug here and you combine it with the RemoveAll.... you may wipe your entire disk.

So I refactor this to take the root of the schema as the base. 

Also I added the clean flag again for the same reason, I don't want a bug to accidentally wipe the entire disk, it is something I want people to execute carefully. If possible, only on CI

PatrickVienne replied:
thanks for your improvements

PatrickVienne commented:
@MetalBlueberry 
frames are removed. back to you

MetalBlueberry commented:
Alright, I made a few tweaks. 
- I removed the changes to the offline package, you can submitt that on a different PR 
- Removed some information from the read me, as I feel that information doesn't belong there. It should be obvious from looking at the code structure. If you still feel that information is useful, it can be added on a different location as specific documentation.
- Fixed the offline package to have the right import paths and CDN
- Fixed other small things

If you are happy, I'm happy with this. We could merge it

PatrickVienne commented on generator/renderer.go:
```diff
@@ -111,13 +111,19 @@ const generationTemplate = `
-		Tag      string
-	}{CDN: cdnUrl, Template: fmt.Sprintf("fmt.Sprintf(`%s`, head)", generationTemplate), Tag: tag}
+		CDN                    string
+		Template               string
```
`Template`, `Tag`, not needed anymore -> Remove

PatrickVienne commented:
Thanks. One small cleanup comment, and then let's merge it.

MetalBlueberry commented:
Let's Go!!
